File: 309-best-time-to-buy-and-sell-stock-with-cooldown/best-time-to-buy-and-sell-stock-with-cooldown.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        # state:
        # 1 = buying
        # 2 = selling
        # 3 = cooling down
        n = len(prices)
        @cache
        def recur(index:int, state:int) -> int:
            if index == n:
                return 0
            if state == 1:
                sol = 0
                sol = max(sol, - prices[index] + recur(index + 1, 2), recur(index + 1, 1))
                return sol
            if state == 2:
                sol = 0
                sol = max(sol,prices[index] + recur(index+1, 3), recur(index+1, 2))
                return sol
            return recur(index+1, 1)
        return recur(0, 1)

File: 309-best-time-to-buy-and-sell-stock-with-cooldown/best-time-to-buy-and-sell-stock-with-cooldown.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        vector<vector<int>> dp(n+1,vector<int>(3, -1));
        dp[1][0] = -prices[1-1];
        dp[1][1] = -1;
        dp[1][2] = 0;

        for(int i = 2; i <= n;i++) {
            dp[i][0] = max(dp[i-1][0], dp[i-1][2] - prices[i-1]);
            dp[i][1] = max(dp[i-1][1], dp[i-1][0] + prices[i-1]);
            dp[i][2] = max(dp[i-1][2], dp[i-1][1]);
        }

        return max(dp[n][1], dp[n][2]);
    }
};